types/node/url.d.ts

```diff
@@ -869,17 +869,21 @@ declare module 'url' {
          * https://nodejs.org/api/url.html#the-whatwg-url-api
          * @since v10.0.0
          */
-        var URL: typeof globalThis extends { webkitURL: infer URL } ? URL : typeof _URL;
+        var URL:
+            // For compatibility with "dom" and "webworker" URL declarations
+            typeof globalThis extends { onmessage: infer O, URL: infer URL }
```

Comment:
Isn't this circular? Declaring `var URL` in a `global` augmentation means that `globalThis` will have a `URL` even if it isn't defined elsewhere, so you might as well just be testing for `{ onmessage: any }`. The previous check tested the presence of `webkitURL`, which would not be circular.

Also, why are you using `infer O`? The resulting type variable isn't used anywhere else.

Reply:
1) Seems that there's no recursion for `URL` because the `onmessage` type check goes first. Should I replace the current one with smth like?
```ts
var URL:
  // For compatibility with "dom" and "webworker" URL declarations
  typeof globalThis extends { onmessage: any }
    ? typeof globalThis.URL
    : typeof _URL;
```

2) removed redundant `infer O`

Reply:
@rbuckton

Reply:
`typeof globalThis.URL` is still circular, because it just references the `URL` you've defined if nothing else is present. The previous code could use `webkitURL` because it was a *different* declaration. Unfortunately, `lib.webworker.generated.d.ts` doesn't have a different definition of `URL` to infer from, though maybe you could do something like this:

```ts
var URL: 
  // for compatibility with "dom" URL
  typeof globalThis extends { webkitURL: infer URL } ? URL :
  // for compatibility with "webworker" URL
  typeof globalThis extends { WebSocket: new (url: string | infer URL) => any } ? URL :
  typeof _URL;
```

Though the `WebSocket` variant could work for both dom and webworker.

Reply:
`typeof globalThis.URL is still circular`, but why this does not raise an error (`error TS2502: 'URL' is referenced directly or indirectly in its own type annotation.`)? Doesn't that mean, it refers to the `URL` defined by `lib.dom/lib.webworker`?

Reply:
Honestly, I'm surprised it's not an error. If you remove `"dom"` from the `"lib"` section in the tsconfig and change the type to `var URL: typeof globalThis.URL` then it *is* an error, but the conditional type keeps that from happening.